token/program-2022/src/extension/reallocate.rs

```diff
@@ -71,20 +73,42 @@ pub fn process_reallocate(
 
     // if additional lamports needed to remain rent-exempt, transfer them
     let rent = Rent::get()?;
-    let new_minimum_balance = rent.minimum_balance(needed_account_len);
-    let lamports_diff = new_minimum_balance.saturating_sub(token_account_info.lamports());
-    invoke(
-        &system_instruction::transfer(payer_info.key, token_account_info.key, lamports_diff),
-        &[
-            payer_info.clone(),
-            token_account_info.clone(),
-            system_program_info.clone(),
-        ],
-    )?;
+    let new_rent_exempt_reserve = rent.minimum_balance(needed_account_len);
 
-    // unpack to set account_type, if needed
-    let mut token_account = token_account_info.data.borrow_mut();
-    set_account_type::<Account>(&mut token_account)?;
+    let current_lamport_reserve = token_account_info
+        .lamports()
+        .checked_sub(native_token_amount.unwrap_or(0))
+        .ok_or(TokenError::Overflow)?;
+    let lamports_diff = new_rent_exempt_reserve.saturating_sub(current_lamport_reserve);
+    if lamports_diff > 0 {
```

Comment:
after some long thinking, i concluded conditionally omitting `invoke(transfer)` is safe to save CUs. this allows anything to be passed in as `payer_info` and `system_program_info` if reallocated account is funded adequately, while not documented so.

That safeness is based on the fact callers shouldn't rely on signedness of `payer` indirectly via internal workings of `reallocate` in this specific IX use case. In contract, `system_program::transfer` do check full-validation even for the 0-lamport transfer case.

Reply:
That was my thinking too, I'm glad that you agree!